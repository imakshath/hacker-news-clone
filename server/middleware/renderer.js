import React from 'react'
import ReactDOMServer from 'react-dom/server'
import { Provider as ReduxProvider } from 'react-redux';
// import our main App component
import App from '../../src/App';

const path = require("path");
const fs = require("fs");

export default (store) => (req, res, next) => {
    // point to the html file created by CRA's build tool
    const filePath = path.resolve(__dirname, '..', '..', 'build', 'index.html');

    fs.readFile(filePath, 'utf8', (err, htmlData) => {
        if (err) {
            return res.status(404).end()
        }

        const html = ReactDOMServer.renderToString(
            <ReduxProvider store={store}>
                <App />
            </ReduxProvider>
        );

        // inject the rendered app into our html and send it
        return res.send(
            htmlData.replace(
                '<div id="root"></div>',
                `<div id="root">${html}</div>`
            )
        );
    });
}