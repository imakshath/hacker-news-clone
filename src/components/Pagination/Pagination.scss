
@import '../../styles/variables';

.pagination-wrapper {
    display: flex;
    justify-content: flex-end;
    align-items: center;
    color: $base-bg-color;
    padding: 5px 20px 5px 10px;
}

.vertical-line {
    border-right: 2px solid $base-bg-color;
    height: 13px;
    margin: 2px 5px 0 5px;
}