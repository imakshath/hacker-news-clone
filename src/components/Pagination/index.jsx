import React from 'react';
import { Link } from '@reach/router';
import './Pagination.scss';

function Pagination({ page, isFetching, totalPages }) {
    const isPrevDisabled = React.useMemo(() => {
        return ((page > 1) && !isFetching) ? false : true
    }, [page, isFetching]);

    const isNextDisabled = React.useMemo(() => {
        return ((page < totalPages) && !isFetching) ? false : true
    }, [page, totalPages, isFetching]);

    return (
        <div className="pagination-wrapper" data-testid="pagination-wrapper">
            <Link
                disabled={isPrevDisabled}
                className="text-bold base-bg-color"
                to={`/?page=${page - 1}`}
            >
                Previous
            </Link>
            <span className="vertical-line base-bg-color"></span>
            <Link
                disabled={isNextDisabled}
                className="text-bold base-bg-color"
                to={`/?page=${page + 1}`}
            >
                Next
            </Link>
        </div>
    );
}

export default React.memo(Pagination);