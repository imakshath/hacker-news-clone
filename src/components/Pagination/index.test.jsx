import React from 'react';
import { render, cleanup } from '@testing-library/react';
import renderer from 'react-test-renderer';
import Pagination from '.';

afterEach(cleanup);

describe('<Pagination /> component', () => {
    let wrapper;
    const props = {
        page: 1,
        isFetching: false,
        totalPages: 10,
    };
    
    beforeEach(() => {
        wrapper = render(
            <Pagination {...props} />
        );
    });

    test('should render without crash', () => {
        expect(wrapper.container).toBeDefined();
    });

    test('should get by testid', () => {
        expect(wrapper.getByTestId('pagination-wrapper')).toBeDefined();
    });

    test('should match snapshot', () => {
        const tree = renderer.create(<Pagination />).toJSON();
        expect(tree).toMatchSnapshot();
    });
    
});