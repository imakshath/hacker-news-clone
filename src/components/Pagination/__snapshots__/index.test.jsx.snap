// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> component should match snapshot 1`] = `
<div
  className="pagination-wrapper"
  data-testid="pagination-wrapper"
>
  <a
    className="text-bold base-bg-color"
    disabled={true}
    href="/?page=NaN"
    onClick={[Function]}
  >
    Previous
  </a>
  <span
    className="vertical-line base-bg-color"
  />
  <a
    className="text-bold base-bg-color"
    disabled={true}
    href="/?page=NaN"
    onClick={[Function]}
  >
    Next
  </a>
</div>
`;
